subsys/net/lib/lwm2m/lwm2m_rw_cbor.c

```diff
@@ -0,0 +1,559 @@
+			goto tstring;
+		case CBOR_TAG_TIME_NUM:
+			goto tnumerical;
+		default:
+			LOG_WRN("expected tagged date/time, got tag %" PRIu32 "", tag);
+			return -EBADMSG;
+		}
+	} else { /* TODO: Tags are optional, right? */
+		;
+	}
+
+tstring:
+	if (!tstrx_decode(states, &hndl)) {
+		/* Might be numerical */
+		goto tnumerical;
+	}
+
+	unsigned long ival = strtoul(time_str, &end, 0);
+
+	errno = 0;
+
+	if (errno == ERANGE) {
+		LOG_WRN("unable to decode a date/time string");
+		errno = 0;
+		return -EBADMSG;
+	}
+
+	*value = ival;
+
+	goto advance;
+
+tnumerical:
+	if (!uintx32_decode(states, (uint32_t *)value)) {
+		LOG_WRN("unable to decode seconds since Epoch");
+		return -EBADMSG;
+	}
+
+advance:
+	data_len = ICTX_CBOR_R_SZ(states[0].payload, in);
+	in->offset += data_len;
+
+	return cbor_tag_sz + data_len;
+}
+
+static size_t get_bool(struct lwm2m_input_context *in, bool *value)
+{
+	cbor_state_t states[1];
+
+	new_state(states, 1, ICTX_BUF_R_REGION(in), 1);
+
+	if (!boolx_decode(states, value)) {
+		LOG_WRN("unable to decode a boolean value");
+		return -EBADMSG;
+	}
+
+	int len = ICTX_CBOR_R_SZ(states[0].payload, in);
+
+	in->offset += len;
+
+	return len;
+}
+
+static size_t get_opaque(struct lwm2m_input_context *in, uint8_t *value, size_t buflen,
+			 struct lwm2m_opaque_context *opaque, bool *last_block)
+{
+	cbor_state_t states[1];
+	cbor_string_type_t info = { 0 };
+	int ret;
+
+	/* Get the CBOR header only on first read. */
+	if (opaque->remaining == 0) {
+		/* FIXME: When data is spread between non-contiguous blocks this range setting will
+		 * break the whole thing. Luckily we get the data length out though.
+		 */
+		new_state(states, 1, ICTX_BUF_R_REGION(in), 1);
+
+		ret = bstrx_cbor_start_decode(states, &info);
+
+		/* FIXME: Uncomment once the range check is working */
+		(void)ret;
+		/*
+		 * if(!ret) {
+		 *	LOG_WRN("unable to decode opaque data header");
+		 *	return -EBADMSG;
+		 * }
+		 */
+
+		opaque->len = info.len;
+		opaque->remaining = info.len;
+
+		/* FIXME: Workaround for the bad range check.
+		 * Now it's CBOR data item header plus possible extended count
+		 */
+		if (info.len == 0) {
+			LOG_WRN("unable to decode opaque data header");
+			return -EBADMSG;
+		} else if (info.len <= 23) {
+			in->offset += 1;
+		} else if (info.len <= UINT8_MAX) {
+			in->offset += 1 + 1;
+		} else if (info.len <= UINT16_MAX) {
+			in->offset += 1 + 2;
+		} else if (info.len <= UINT32_MAX) {
+			in->offset += 1 + 4;
+		}
+	}
+
+	return lwm2m_engine_get_opaque_more(in, value, buflen, opaque, last_block);
+}
+
+static size_t get_objlnk(struct lwm2m_input_context *in, struct lwm2m_objlnk *value)
+{
+	char objlnk[sizeof("65535:65535")] = { 0 };
+	char *end;
+	char *idp = objlnk;
+
+	value->obj_id = LWM2M_OBJLNK_MAX_ID;
+	value->obj_inst = LWM2M_OBJLNK_MAX_ID;
+
+	int len = get_string(in, objlnk, sizeof(objlnk) - 1);
+
+	errno = 0;
+
+	for (int idx = 0; idx < 2; idx++) {
+		unsigned long id = strtoul(idp, &end, 10);
```

Comment:
This is nitpicking, but to be obviously pedantic, I would follow See CERT-C [ERR30-C](https://wiki.sei.cmu.edu/confluence/pages/viewpage.action?pageId=87152351)
So set `errno = 0;` just before the call, then check the return value as well.

Now you have to look twice, to see if errno is ever cleared after someone sets it.

Reply:
You're correct. To be fixed.

Reply:
Fixed.